packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts

```diff
@@ -137,12 +137,13 @@ export class DependencyManager {
     );
   }
 
-  async addProvider(
-    constraint: ProviderConstraint
-  ): Promise<{ addedLocalProvider: boolean }> {
+  async addProvider(constraint: ProviderConstraint): Promise<{
```

Comment:
nit: The method is a bit leaky now, as it either has no side-effect (when the provider ends up being available as pre-built provider) or it does have a side-effect (when it's added locally to the `cdktf.json`).
I could see us changing the method name to try to reflect the non-side-effect nature or to maybe refactor this in a way that changes to the local-adding bit to have no side-effect, too, and also return "installation instructions" aka a constraint that can be added later on.

Reply:
While it is pretty rare – in theory (e.g. with a slow filesystem) reading the local providers, filtering them, adding another local provider, and writing them might be a different race condition ([source](https://github.com/hashicorp/terraform-cdk/blob/950c919b901157673f5afda47dc0e4b7ecd3f231/packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts#L366)). Maybe refactoring that bit in a similar way to the pre-built provider packages would get rid of that race condition while solving my nit about this inconsistency / about the methods name 🤔